### PR TITLE
List target backends registered by plugins in C API.

### PR DESCRIPTION
**Pull request by ScottTodd**
Follow-up to https://github.com/openxla/iree/pull/13262

Plugins may register target backends, so the global registry does not reliably enumerate all targets.

The `--iree-hal-list-target-backends` flag callback runs globally, before session/invocation initialization and plugin activation, so it was missing the `cuda` backend in its list. I've fixed that by creating an internal/ephemeral `Session`.

---

Also added a new `--iree-list-plugins` CLI flag.

### CODE REVIEW

**Comment by ScottTodd** [comment]
Ping?

File: compiler/bindings/c/iree/compiler/embedding_api.h [hunk 1]
```diff
@@ -117,10 +117,6 @@ IREE_EMBED_EXPORTED void ireeCompilerSetupGlobalCL(int argc, const char **argv,
 // This must be called prior to library unloading.
 IREE_EMBED_EXPORTED void ireeCompilerGlobalShutdown();
 
-// Invokes a callback with each registered HAL target backend.
```
Comment (stellaraccident) [inline compiler/bindings/c/iree/compiler/embedding_api.h:120]: This is a versioned API. At a minimum, we need to keep this entry-point as a no-op and add the new one with a new version. See how it is done for ireeCompilerEnumeratePlugins.

However, this API is really just a kludge. I think I would prefer that it just internally create an invocation/session/activate plugins and then enumerate. We can update the comment to indicate that it does so and is really only suitable for global CLI-like tools.

Comment (ScottTodd) [reply compiler/bindings/c/iree/compiler/embedding_api.h:120]: Gotcha. I reverted the breaking changes and moved the session creation and plugin activation into the existing API implementation. PTAL.

**Review by stellaraccident — APPROVED** [review]
Thanks!

**Pull request merged**